Gutenlypso: Disable the Publish button if the email is unverified

mmtr opened the pull request:
#### Changes proposed in this Pull Request

Fixes #29539.

This PR modifies the `PostPublishButtonOrToggle` component from the `edit-post` in order to disable the Publish button if the email of the user is not verified.

Since it is now very different from the original component in the `@wordpress/edit-post` package, I decided to move it `/gutenberg/editor/components` so it's clear we're using a customized version.

Apart from displaying the Publish button, this component now do 3 more things if the email is unverified:
* Dispatches a `lockPostSaving` action, so the Publish button is disabled. Note that if the Pre-publish checks are enabled, only the Publish button from the Pre-publish sidebar will be disabled.
* Displays a warning notice with the same messages we display on the classic editor ([handled by `EditorGroundControl`](https://github.com/Automattic/wp-calypso/blob/2964f6c86eafb1a113531d68e60215bd32644268/client/post-editor/editor-ground-control/index.jsx#L77-L92)).
* Renders the `VerifyEmailDialog` component when the "Learn more" link is clicked.

<img width="1274" alt="screen shot 2018-12-19 at 11 53 35" src="https://user-images.githubusercontent.com/1233880/50216290-a37e2e00-0385-11e9-8768-16870ea3080f.png">

<img width="1276" alt="screen_shot_2018-12-19_at_11_53_46" src="https://user-images.githubusercontent.com/1233880/50216277-96613f00-0385-11e9-966f-c999b4a36988.png">

#### Testing instructions

* Go to http://calypso.localhost:3000/start and create a new user without verifying the email. Don't use an @automattic.com email since they need to be always verified before logging in. You can use a personal Gmail account, even if it's already registered, by appending a `+` suffix (i.e. if your email is example@gmail.com, you can use example+20181219test@gmail.com). WordPress.com will considere it a different email but example@gmail.com will still receive any email sent to example+20181219test@gmail.com).
* Start creating a new post (it should use Gutenlypso by default).
* Make sure that a warning notice appears with the message "To publish, check your email and confirm your address. Learn more"
<img width="453" alt="screen shot 2018-12-19 at 11 12 04" src="https://user-images.githubusercontent.com/1233880/50216797-0de39e00-0387-11e9-82b9-839d3c0e4a03.png">

* Click on Learn more and check that the below dialog appears.
<img width="560" alt="screen_shot_2018-12-19_at_11_53_46" src="https://user-images.githubusercontent.com/1233880/50216878-500cdf80-0387-11e9-87af-4d8251b961dc.png">

* Confirm that the "Publish..." button is not disabled and that after clicking on it, the "Publish" button on the pre-publish sidebar is disabled and you cannot click on it.
<img width="1276" alt="screen shot 2018-12-19 at 12 14 23" src="https://user-images.githubusercontent.com/1233880/50217097-0bce0f00-0388-11e9-85c1-9d387e94c1cc.png">
<img width="1275" alt="screen shot 2018-12-19 at 12 16 33" src="https://user-images.githubusercontent.com/1233880/50217099-0bce0f00-0388-11e9-8240-70306ba59c30.png">

* Change the publish date to some date in the future and make sure that the verification email notice changes to "To schedule, check your email and confirm your address."
![image](https://user-images.githubusercontent.com/1233880/50217395-d4139700-0388-11e9-887c-e9570fc12433.png)

* On the More menu, click on Options and uncheck the "Enable Pre-publish Checks" option.
<img width="419" alt="screen shot 2018-12-19 at 12 23 55" src="https://user-images.githubusercontent.com/1233880/50217451-ff968180-0388-11e9-9acf-39f9e5ef8c4a.png">

* Note that the "Schedule..." button in the header changes to "Schedule" and it is now disabled and you cannot click on it.
<img width="1277" alt="screen shot 2018-12-19 at 12 25 12" src="https://user-images.githubusercontent.com/1233880/50217524-2eacf300-0389-11e9-9e23-c4eb11c03bd7.png">

* Verify now the email, reload the editor and check that no notice is displayed and that the Publish/Schedule buttons are enabled.

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,223 @@
+	applyWithDispatch,
+	withViewportMatch( { isLessThanMediumViewport: '< medium' } ),
+	withState( {
+		showEmailVerificationNotice: false,
```
What would be the advantage of using `withState` instead of `this.state`?

mmtr replied:
Actually, none on this particular component. 

I think `withState` is great for functional components, because it allows to have state on them, without needing to transform them to classes. That's why I like to use `withState`.

But this component is already a class component so `withState` is not adding any specific benefit. 

Do you think I should use `this.state`?

Copons replied:
That's interesting, I didn't really know of `withState`.

To be honest, though, in Calypso we are perfectly fine in using the classical component state, and `withState` feels a bit out of place to me.

Also, as I said [elsewhere](https://github.com/Automattic/wp-calypso/pull/29594/files#r242898999), please change this variable name because it's rather confusing. 🙂

```diff
- showEmailVerificationNotice
+ showEmailVerificationDialog
```

mmtr commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,223 @@
+				actions: [
+					{
+						label: translate( 'Learn More' ),
+						url: true, // This causes a warning because true is not a valid href attribute, but it is
```
Added a work-around in 2441771 by using `#` as href value. Props to @Copons !

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,225 @@
+	}
+}
+
+const applyWithSelect = withSelect( select => {
```
I think this change is unnecessary compared to the original Gutenberg component.

We don't need to translate `verificationNoticeLabel` every time the state changes, but instead we should only render it translated when strictly required.
We usually do this in a component method, e.g.
```jsx
getVerificationNoticeLabel = () => {
  const { isPublished, isBeingScheduled, translate } = this.props;
  if ( isPublished ) return translate( '...' );
  else if ( isBeingScheduled ) return translate( '...' );
  return translate( '...' );
};
```

This means that in `componentDidUpdate` you should be more precise in when calling `this.togglePostSaving();`

As a side note, you might have notice that I'm suggesting to use `translate` as a prop.
Such prop is created by the `localize()` higher order component, and it's the [suggested way of using i18n-calypso](https://github.com/Automattic/i18n-calypso/blob/9556007bd2dd6d3bb88edfd0c5e8e4275edf06b4/README.md).

Copons replied:
To be more clear, I'd move the notice label localization into a component method, and restore the original `withSelect` altogether:

```jsx

	withSelect( select => ( {
		hasPublishAction: get(
			select( 'core/editor' ).getCurrentPost(),
			[ '_links', 'wp:action-publish' ],
			false
		),
		isBeingScheduled: select( 'core/editor' ).isEditedPostBeingScheduled(),
		isPending: select( 'core/editor' ).isCurrentPostPending(),
		isPublished: select( 'core/editor' ).isCurrentPostPublished(),
		isPublishSidebarEnabled: select( 'core/editor' ).isPublishSidebarEnabled(),
		isPublishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
		isScheduled: select( 'core/editor' ).isCurrentPostScheduled(),
	} ) ),
```

mmtr replied:
In fact, I had an implementation using this approach in some moment, but decided to change it exactly for what you say about `componentDidUpdate`:

> This means that in `componentDidUpdate` you should be more precise in when calling `this.togglePostSaving();``

It ended up being a bit more complex, checking 3 props (`userNeedsVerification`, `isPublished` and `isBeingScheduled`) instead of 2 (`userNeedsVerification` and `verificationNoticeLabel`).

But this is actually making more complex the `withSelect`, so it makes sense to revert it since that will also make possible to use the `localize` HOC.

Done in d5d6164.

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,225 @@
+ */
+import React, { Component, Fragment } from 'react';
+import { get } from 'lodash';
+import { translate } from 'i18n-calypso';
```
Mentioned already in https://github.com/Automattic/wp-calypso/pull/29594/files#r242893520, but for clarity:

The [recommended way](https://github.com/Automattic/i18n-calypso/blob/9556007bd2dd6d3bb88edfd0c5e8e4275edf06b4/README.md) of using `i18n-calypso` is through the `localize` HOC.

E.g.
```jsx
import { localize } from 'i18n-calypso';

const FooBar = ( { translate } ) => ( <div>{ translate( 'Howdy!' ) }</div> );

export localize( FooBar );
```

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,225 @@
+		const { userNeedsVerification, verificationNoticeLabel } = this.props;
+		if (
+			userNeedsVerification !== prevProps.userNeedsVerification ||
+			verificationNoticeLabel !== prevProps.verificationNoticeLabel
```
I'm wondering why do we need to check if `verificationNoticeLabel` changed—or, in other words, [if the post "status" changed](https://github.com/Automattic/wp-calypso/pull/29594/files#diff-f4811af222e472a6d2de60560f7624b7R169) between `isPublished`, `isBeingScheduled`, or otherwise.

Wouldn't be enough for our purpose to simply disable saving when `userNeedsVerification`, and enable it when the user doesn't?

mmtr replied:
It's needed because `togglePostSaving` (to be changed to `toggleLockPostSaving`) does 2 things:
* Locks/unlocks the post saving
* Creates a notice with a message which is based on the post status

If we don't check the post status, the notice won't change from "To publish, check your email and confirm your address." to "To schedule, check your email and confirm your address." when scheduling the post.

Copons replied:
Makes sense 👍

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,225 @@
+					isToggle={ component === IS_TOGGLE }
+					onToggle={ togglePublishSidebar }
+				/>
+				{ showEmailVerificationNotice && (
```
Since we have an actual email verification notice in this component, I'd definitely change this into `showEmailVerificationDialog` to avoid confusion.

Copons commented:
I've left a lot of comments, but I think this is definitely on the right track @mmtr! ✨ 

---

There is one other thing that is bugging me A LOT, and it's unrelated to this change, but this change made me realize it.

Basically, as of https://github.com/WordPress/gutenberg/pull/11543 (cc @nosolosw), `PostPublishButton` doesn't have a `disabled` attribute anymore, but uses `aria-disabled` instead.
This seems to fix the focus loss issue they're having, but also... I can click on the button when `aria-disabled` and save the post (recently, I've happened to do this countless times with the disabled Update button).

In this particular case, you can try clicking on Publish, which simply opens the sidebar (it shouldn't do that, should it?).
Then click on it again and, even if the publish will silently fail (for any number of reasons, unverified email, empty content, etc), you can observe in the sidebar the Publish time changing from "immediately" to the actual current datetime.

I think that we should at the very least tackle this via CSS.
E.g.
```css
.components-button.is-primary[aria-disabled="true"] {
  pointer-events: none;
}
```

mmtr commented:
Thanks for the review @Copons! I just finished to push all the commits addressing your comments.

> Basically, as of WordPress/gutenberg#11543 (cc @nosolosw), PostPublishButton doesn't have a disabled attribute anymore, but uses aria-disabled instead.

Good catch! In fact, this bug is present in Core, so I filed a bug there: https://github.com/WordPress/gutenberg/issues/13014

In the meantime and since this is unrelated to this PR, I'd say we can fix it in Calypso with a follow-up PR applying your suggestion of using `pointer-events: none;`.

> In this particular case, you can try clicking on Publish, which simply opens the sidebar (it shouldn't do that, should it?).

Only if the content is empty. If the email is not verified, only the button in the sidebar will be disabled. This is because `PostPublishButton` uses the `isPostSavingLocked` selector (toggled by dispatching the `lockPostSaving`/`unlockPostSaving` actions) for disabling the "Publish" button only if the Pre-publish checks are deactivated (in other words, if `PostPublishButton` is rendered with `isToggle` equals to `false`).

Copons commented on client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx:
```diff
@@ -0,0 +1,217 @@
+		const userNeedsVerification =
+			! isCurrentUserEmailVerified( state ) &&
+			! isVipSite( state, siteId ) &&
+			! isUnlaunchedSite( state, siteId );
```
Testing the PR I've got into a funny case that somehow "breaks" these checks.

- Register new user.
- Keep the email unverified.
- Create a new post.
- Publish it: there are no unverified email warnings, as the site is "unlaunched".

This is correct: we allow unverified users to publish, as long as their site is not yet public.

Though, to launch a site, the user first needs to verify their email.

So, I wonder: when can it happen that a user has both a verified email and an unlaunched site—or viceversa?
Those two cases would be the only two that would enable the whole `userNeedsVerification` flow, with the warning and the dialog, etc. (in both Gutenberg and Classic)

cc @scruffian for adding `isUnlaunchedSite` to the equivalent of this in `EditorGroundControl` in #27962.

scruffian replied:
It's possible for a user to have an unlaunched site and a verified email address. Since the process to launch a site would be to first verify the email address and then launch the site, any users who are part way through that process will be in that state.

The reverse state should be impossible - we should never allow users with unverified email addresses to launch a site.

Does that answer your question?

Copons replied:
Absolutely yes, thank you very much @scruffian!
I didn't proceed with the test ~because I'm that lazy~ to keep the user email unverified, and somehow just assumed verifying the email would also automatically launch (which would definitely be odd).

Copons replied:
(Sorry for the delay from my previous answer, I just now got to test it)

So, what happens then is that:
- Verified user can have launched and unlaunched sites.
- Unverified user can only have unlaunched sites.

In other words, this would never be true:
```es6
const userNeedsVerification =
	! isCurrentUserEmailVerified( state ) &&
	! isUnlaunchedSite( state, siteId );
```

In other words, we don't need the unverified email notice and the logic that prevents publishing a post, for both Guten and Classic.

Am I wrong on this?
Like, it's almost EOD and I've got the flu, I might very well have missed something. 😄

mmtr replied:
> Unverified user can only have unlaunched sites.

I think this is not true. According to #27962, only private sites are unlaunched, so an unverified user can have a launched site if it is public.

kwight commented:
I noticed that while a new post loads with the verification notice and disabled Publish button, as soon as I enter content, the button is enabled again:

<img width="1127" alt="screen shot 2018-12-19 at 8 58 44 am" src="https://user-images.githubusercontent.com/349751/50235256-5346a200-036c-11e9-85fc-14e73aa046b0.png">

The second Publish button will still be disabled, but like @Copons noticed, it's functional.

Feel free to ignore this, or we can consider for a follow up: I don't think this is the right place to handle this. It would be better as part of `/me/account`, which should really have all of this regardless. If that screen handled this dialog content and actions (it could even be inline, rather than needing a dialog), _any_ part of Calypso blocked by unverified emails could just link to `/me/account` (which our "Learn more" link would do with this approach). Of course, this ends up needing design input and so forth, so kicking down the road may be the better choice today.

mmtr commented:
@kwight as I mention on https://github.com/Automattic/wp-calypso/pull/29594#issuecomment-448619525, it's expected to have the first "Publish..." button enabled if the "Pre-publish checks" are activated.

You can check if they are enabled by opening the "More" menu and clicking on "Options":
<img width="419" alt="screen shot 2018-12-19 at 12 23 55" src="https://user-images.githubusercontent.com/1233880/50217451-ff968180-0388-11e9-9acf-39f9e5ef8c4a.png">

So, to recap:
- If the "Pre-publish checks" are activated, Gutenberg doesn't disable the first "Publish..." button because it is not considered a button but a toggle. The toggle will open a sidebar with the actual "Publish" button which is the one Gutenberg disables when the `lockPostSaving` is dispatched.
- If the "Pre-publish checks" are deactivated, Gutenberg disables the first and only "Publish" button in the header because there isn't any toggle than open any sidebar. That button is the actual button that will publish the post affected by the `lockPostSaving` action.

mmtr commented:
I changed my mind and I think we should address here the issue that is allowing to click on the disabled Publish button, since it's very related to this PR so I added the needed CSS in 492ab33.

Once the issue is solved in Core, we can revert it.
